File: src/main/java/com/logics/logics/controllers/GameRoomController.java
package com.logics.logics.controllers;


import com.logics.logics.entities.GameRoom;
import com.logics.logics.services.GameRoomService;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@RestController
@RequestMapping("/api/rooms")
public class GameRoomController {

  private final GameRoomService gameRoomService;

  public GameRoomController(GameRoomService gameRoomService) {
    this.gameRoomService = gameRoomService;
  }

  @PostMapping
  public Mono<ResponseEntity<GameRoom>> createRoom(@AuthenticationPrincipal UserDetails userDetails,
      @RequestParam String name,
      @RequestParam int maxPlayers,
      @RequestParam String category) {
    return gameRoomService.createRoom(userDetails.getUsername(), name, maxPlayers, category)
        .map(ResponseEntity::ok);
  }

  @GetMapping
  public Flux<GameRoom> getAvailableRooms() {
    return gameRoomService.getAvailableRooms();
  }

  @PostMapping("/{roomId}/join")
  public Mono<ResponseEntity<GameRoom>> joinRoom(@AuthenticationPrincipal UserDetails userDetails,
      @PathVariable String roomId) {
    return gameRoomService.joinRoom(roomId, userDetails.getUsername())
        .map(ResponseEntity::ok)
        .onErrorResume(e -> Mono.just(ResponseEntity.badRequest().build()));
  }

  @PostMapping("/{roomId}/leave")
  public Mono<ResponseEntity<Void>> leaveRoom(@AuthenticationPrincipal UserDetails userDetails,
      @PathVariable String roomId) {
    return gameRoomService.leaveRoom(roomId, userDetails.getUsername())
        .then(Mono.just(ResponseEntity.ok().<Void>build()));
  }

  @PostMapping("/{roomId}/start")
  public Mono<ResponseEntity<GameRoom>> startGame(@PathVariable String roomId) {
    return gameRoomService.startGame(roomId)
        .map(ResponseEntity::ok)
        .onErrorResume(e -> Mono.just(ResponseEntity.badRequest().build()));
  }

  @PostMapping("/{roomId}/disband")
  public Mono<ResponseEntity<Void>> disbandRoom(@PathVariable String roomId) {
    return gameRoomService.disbandRoom(roomId)
        .then(Mono.just(ResponseEntity.ok().<Void>build()));
  }
}

File: src/main/java/com/logics/logics/controllers/GameRoomController.kt
package com.logics.logics.controllers

import com.logics.logics.dto.GameRoomRequest
import com.logics.logics.entities.GameRoom
import com.logics.logics.services.GameRoomService
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.web.bind.annotation.*
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@RestController
@RequestMapping("/api/rooms")
class GameRoomController(private val gameRoomService: GameRoomService) {

    private val logger = LoggerFactory.getLogger(GameRoomController::class.java)

    @PostMapping
    fun createRoom(
        @AuthenticationPrincipal userDetails: UserDetails,
        @RequestBody request: GameRoomRequest
    ): Mono<ResponseEntity<GameRoom>> {
        return gameRoomService.createRoom(
            userDetails.username,
            request.name ?: "Комната без названия",
            request.maxPlayers,
            request.category ?: "Без категории"
        ).map { ResponseEntity.ok(it) }
            .doOnError { e -> logger.error("Ошибка при создании комнаты: {}", e.message) }
            .onErrorResume { Mono.just(ResponseEntity.badRequest().build()) }
    }

    @GetMapping
    fun getAvailableRooms(): Flux<GameRoom> {
        return gameRoomService.getAvailableRooms()
    }


    @GetMapping("/{roomId}")
    fun getRoomById(@PathVariable roomId: Long): Mono<ResponseEntity<GameRoom>> {
        return gameRoomService.findById(roomId)
            .map { room -> ResponseEntity.ok(room) }
            .defaultIfEmpty(ResponseEntity.notFound().build())  // Возвращаем 404, если комната не найдена
    }

  }